test_shaders.py

```diff
@@ -410,26 +484,50 @@ def test_shader_hlsl(stats, shader, update, keep, opt):
     regression_check(shader, hlsl, update, keep, opt)
     remove_file(spirv)
 
-def test_shaders_helper(stats, shader_dir, update, malisc, keep, opt, backend):
-    for root, dirs, files in os.walk(os.path.join(shader_dir)):
+def test_shader_reflect(stats, shader, update, keep, opt):
+    joined_path = os.path.join(shader[0], shader[1])
+    print('Testing shader reflection:', joined_path)
+    is_spirv = shader_is_spirv(shader[1])
+    noopt = shader_is_noopt(shader[1])
+    spirv, reflect = cross_compile_reflect(joined_path, is_spirv, opt and (not noopt))
+    regression_check_reflect(shader, reflect, update, keep, opt)
+    remove_file(spirv)
+
+def test_shader_file(relpath, stats, shader_dir, update, keep, opt, backend):
+    if backend == 'msl':
+        test_shader_msl(stats, (shader_dir, relpath), update, keep, opt)
+    elif backend == 'hlsl':
+        test_shader_hlsl(stats, (shader_dir, relpath), update, keep, opt)
+    elif backend == 'reflect':
+        test_shader_reflect(stats, (shader_dir, relpath), update, keep, opt)
+    else:
+        test_shader(stats, (shader_dir, relpath), update, keep, opt)
+
+def test_shaders_helper(stats):
+    all_files = []
+    for root, dirs, files in os.walk(os.path.join(args.folder)):
         files = [ f for f in files if not f.startswith(".") ]   #ignore system files (esp OSX)
         for i in files:
             path = os.path.join(root, i)
-            relpath = os.path.relpath(path, shader_dir)
-            if backend == 'msl':
-                test_shader_msl(stats, (shader_dir, relpath), update, keep, opt)
-            elif backend == 'hlsl':
-                test_shader_hlsl(stats, (shader_dir, relpath), update, keep, opt)
-            else:
-                test_shader(stats, (shader_dir, relpath), update, keep, opt)
+            relpath = os.path.relpath(path, args.folder)
+            all_files.append(relpath)
+
+    # The child processes in parallel execution mode don't have the proper state for the global args variable, so 
+    # at this point we need to switch to explicit arguments
+    if args.parallel:
+        pool = multiprocessing.Pool(multiprocessing.cpu_count())
```

Comment:
Cool.

Reply:
Thanks.  I was running into failures of the regression tests when I was refactoring the type_to_glsl and image_type_to_glsl functions, and it was annoying me that each test would take so long.  Had to learn the python mechanism for parallelism and for partial function binding, but it ended up being not too complicated.  

I think the multiprocessing approach here may be overkill though.  Most stuff I saw about python threading led rapidly to dire warnings about the global interpreter lock killing multi-threading performance gains, hence the need to essentially use a variant of `fork` to go multi-process instead of multi-thread.  However, since the vast majority of the time in this script is actually already spent launching other processes, I suspect it would not have been a problem to just use threads.

Reply:
multiprocess makes sense. We just want to invoke completely separate things in parallel.